symengine/fields.cpp

```diff
@@ -0,0 +1,267 @@
+            else
+                dict_quo.insert(dict_quo.end(),{it.first - deg_divisor, it.second});
+        }
+        *quo = GaloisFieldDict(dict_quo, modulo_);
+        *rem = GaloisFieldDict(dict_rem, modulo_);
+    }
+}
+
+GaloisFieldDict GaloisFieldDict::gf_lshift(const integer_class n) const
+{
+    map_uint_mpz dict_out;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        dict_out.insert(dict_out.end(), {iter.first + n_val, iter.second});
+    }
+    return GaloisFieldDict(dict_out, modulo_);
+}
+
+void GaloisFieldDict::gf_rshift(const integer_class n,
+                            const Ptr<GaloisFieldDict> &quo,
+                            const Ptr<GaloisFieldDict> &rem) const
+{
+    map_uint_mpz dict_quo, dict_rem;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        if (iter.first >= n_val)
+            dict_quo.insert(dict_quo.end(), {iter.first - n_val, iter.second});
+        else
+            dict_rem.insert(dict_rem.end(), iter);
+    }
+    *quo = GaloisFieldDict(dict_quo, modulo_);
+    *rem = GaloisFieldDict(dict_rem, modulo_);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_sqr() const
+{
+    return (*this * *this);
+}
+
+GaloisFieldDict GaloisFieldDict::gf_pow(const integer_class n) const
+{
+    if (n == 0) {
+        return GaloisFieldDict({integer_class(1)}, modulo_);
+    }
+    if (n == 1)
+        return static_cast<GaloisFieldDict>(*this);
+    if (n == 2)
+        return gf_sqr();
+    long num = mp_get_si(n);
+    GaloisFieldDict to_sq = static_cast<GaloisFieldDict>(*this);
+    GaloisFieldDict to_ret = GaloisFieldDict({integer_class(1)}, modulo_);
+    while (1) {
+        if (num & 1) {
+            to_ret *= to_sq;
+        }
+        num >>= 1;
+        if (num == 0)
+            return to_ret;
+        to_sq = to_sq.gf_sqr();
+    }
+}
+
+void GaloisFieldDict::gf_monic(integer_class &res,
+                           const Ptr<GaloisFieldDict> &monic) const
+{
+    *monic = static_cast<GaloisFieldDict>(*this);
+    if (dict_.empty()) {
+        res = integer_class(0);
+    } else {
+        res = dict_.rbegin()->second;
+        if (res != integer_class(1)) {
+            integer_class inv, temp;
+            mp_invert(inv, res, modulo_);
+            for (auto &iter : monic->dict_) {
+                temp = inv;
+                temp *= iter.second;
+                mp_fdiv_r(iter.second, temp, modulo_);
+            }
+        }
+    }
+}
+
+GaloisFieldDict GaloisFieldDict::gf_gcd(const GaloisFieldDict &o) const
+{
+    SYMENGINE_ASSERT(modulo_ == o.modulo_);
+    GaloisFieldDict f = static_cast<GaloisFieldDict>(*this);
+    GaloisFieldDict g = o;
+    GaloisFieldDict temp_out;
+    while (not g.dict_.empty()) {
+        f.gf_div(g, outArg(temp_out), outArg(f)); // f, g = f % g, g
+        f.dict_.swap(g.dict_);
+    }
+    integer_class temp_LC;
+    f.gf_monic(temp_LC, outArg(f));
+    return std::move(f);
```

Comment:
Because this will prevent copy-elision. See this answer for more details, http://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value